File: app/schemas/reports.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict


# ============================================================
# DECISION REPORT
# ============================================================

class DecisionReportItem(BaseModel):
    decision_id: int
    title: str
    category: str
    status: str
    created_by: int
    created_date: datetime
    updated_date: Optional[datetime] = None
    alternatives_count: int
    approvals_count: int
    tags: list[str] = []

    model_config = ConfigDict(from_attributes=True)


class DecisionReportSummary(BaseModel):
    total_decisions: int
    draft_decisions: int
    under_review_decisions: int
    approved_decisions: int
    rejected_decisions: int
    archived_decisions: int


class DecisionReportResponse(BaseModel):
    items: list[DecisionReportItem]
    summary: DecisionReportSummary
    page: int
    page_size: int
    total: int


# ============================================================
# APPROVAL REPORT
# ============================================================

class ApprovalReportItem(BaseModel):
    approval_id: int
    decision_id: int
    decision_title: str
    reviewer_id: int
    approval_level: int
    status: str
    assigned_date: datetime
    completed_date: Optional[datetime] = None
    turnaround_days: Optional[float] = None

    model_config = ConfigDict(from_attributes=True)


class ApprovalReportSummary(BaseModel):
    total_approvals: int
    pending_approvals: int
    approved_approvals: int
    rejected_approvals: int
    average_turnaround_days: Optional[float] = None
    completion_rate: float


class ApprovalReportResponse(BaseModel):
    items: list[ApprovalReportItem]
    summary: ApprovalReportSummary
    page: int
    page_size: int
    total: int

# ============================================================
# TEAM REPORT
# ============================================================

class TeamReportItem(BaseModel):
    team_name: str
    members: int
    total_decisions: int
    approved_decisions: int
    rejected_decisions: int
    pending_decisions: int
    approval_rate: float

    model_config = ConfigDict(from_attributes=True)


class TeamReportSummary(BaseModel):
    total_teams: int
    total_members: int
    total_decisions: int
    total_approved: int
    total_rejected: int
    total_pending: int


class TeamReportResponse(BaseModel):
    items: list[TeamReportItem]
    summary: TeamReportSummary
    page: int
    page_size: int
    total: int
# ============================================================
# AUDIT REPORT
# ============================================================

class AuditReportItem(BaseModel):
    audit_id: int
    user_id: Optional[int] = None
    action: str
    entity_type: str
    entity_id: Optional[int] = None
    description: str
    timestamp: datetime
    ip_address: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class AuditReportSummary(BaseModel):
    total_audit_logs: int
    create_actions: int
    update_actions: int
    delete_actions: int
    approve_actions: int
    reject_actions: int
    access_actions: int
    login_actions: int
    logout_actions: int
    submit_actions: int
    archive_actions: int


class AuditReportResponse(BaseModel):
    items: list[AuditReportItem]
    summary: AuditReportSummary
    page: int
    page_size: int
    total: int

File: frontend/src/services/reports.ts
import api from "./api";

export interface DecisionReportItem {
  decision_id: number;
  title: string;
  category?: string;
  status?: string;
  created_by?: number;
  created_date?: string;
  updated_date?: string;
  alternatives_count?: number;
  approvals_count?: number;
  tags?: string[];
}

export interface DecisionReportSummary {
  total_decisions: number;
  draft_decisions: number;
  under_review_decisions: number;
  approved_decisions: number;
  rejected_decisions: number;
  archived_decisions: number;
}

export interface DecisionReportResponse {
  items: DecisionReportItem[];
  summary: DecisionReportSummary;
  page: number;
  page_size: number;
  total: number;
}

export interface ApprovalReportItem {
  approval_id: number;
  decision_id: number;
  decision_title: string;
  reviewer_id: number;
  approval_level: number;
  status: string;
  assigned_date?: string;
  completed_date?: string | null;
  turnaround_days?: number | null;
}

export interface ApprovalReportSummary {
  total_approvals: number;
  pending_approvals: number;
  approved_approvals: number;
  rejected_approvals: number;
  average_turnaround_days?: number | null;
  completion_rate: number;
}

export interface ApprovalReportResponse {
  items: ApprovalReportItem[];
  summary: ApprovalReportSummary;
  page: number;
  page_size: number;
  total: number;
}

export interface TeamReportItem {
  team_name: string;
  members: number;
  total_decisions: number;
  approved_decisions: number;
  rejected_decisions: number;
  pending_decisions: number;
  approval_rate: number;
}

export interface TeamReportSummary {
  total_teams: number;
  total_members: number;
  total_decisions: number;
  total_approved: number;
  total_rejected: number;
  total_pending: number;
}

export interface TeamReportResponse {
  items: TeamReportItem[];
  summary: TeamReportSummary;
  page: number;
  page_size: number;
  total: number;
}

export interface AuditReportItem {
  audit_id: number;
  user_id: number;
  action: string;
  entity_type: string;
  entity_id?: number | null;
  description?: string;
  timestamp?: string;
  ip_address?: string | null;
}

export interface AuditReportSummary {
  total_audit_logs: number;
  create_actions: number;
  update_actions: number;
  delete_actions: number;
  approve_actions: number;
  reject_actions: number;
  access_actions: number;
  login_actions: number;
  logout_actions: number;
  submit_actions: number;
  archive_actions: number;
}

export interface AuditReportResponse {
  items: AuditReportItem[];
  summary: AuditReportSummary;
  page: number;
  page_size: number;
  total: number;
}

export interface ReportFilters {
  category?: string;
  status?: string;
  created_by?: number;
  start_date?: string;
  end_date?: string;
  tag?: string;
  reviewer_id?: number;
  decision_id?: number;
  approval_level?: number;
  team?: string;
  user_id?: number;
  action?: string;
  entity_type?: string;
  entity_id?: number;
  page?: number;
  page_size?: number;
  sort_by?: string;
  sort_order?: string;
}

export async function getDecisionReport(
  params?: ReportFilters,
) {
  const response = await api.get<DecisionReportResponse>(
    "/reports/decisions",
    {
      params,
    },
  );

  return response.data;
}

export async function getApprovalReport(
  params?: ReportFilters,
) {
  const response = await api.get<ApprovalReportResponse>(
    "/reports/approvals",
    {
      params,
    },
  );

  return response.data;
}

export async function getTeamReport(
  params?: ReportFilters,
) {
  const response = await api.get<TeamReportResponse>(
    "/reports/teams",
    {
      params,
    },
  );

  return response.data;
}

export async function getAuditReport(
  params?: ReportFilters,
) {
  const response = await api.get<AuditReportResponse>(
    "/reports/audit",
    {
      params,
    },
  );

  return response.data;
}

function downloadBlob(
  blob: Blob,
  filename: string,
) {
  const url = window.URL.createObjectURL(blob);

  const link = document.createElement("a");
  link.href = url;
  link.download = filename;

  document.body.appendChild(link);
  link.click();

  link.remove();

  window.URL.revokeObjectURL(url);
}

export async function exportDecisionPdf(
  params?: ReportFilters,
) {
  const response = await api.get(
    "/reports/decisions/export/pdf",
    {
      params,
      responseType: "blob",
    },
  );

  downloadBlob(
    response.data,
    "decision_report.pdf",
  );
}

export async function exportDecisionExcel(
  params?: ReportFilters,
) {
  const response = await api.get(
    "/reports/decisions/export/excel",
    {
      params,
      responseType: "blob",
    },
  );

  downloadBlob(
    response.data,
    "decision_report.xlsx",
  );
}

export async function exportApprovalPdf(
  params?: ReportFilters,
) {
  const response = await api.get(
    "/reports/approvals/export/pdf",
    {
      params,
      responseType: "blob",
    },
  );

  downloadBlob(
    response.data,
    "approval_report.pdf",
  );
}

export async function exportApprovalExcel(
  params?: ReportFilters,
) {
  const response = await api.get(
    "/reports/approvals/export/excel",
    {
      params,
      responseType: "blob",
    },
  );

  downloadBlob(
    response.data,
    "approval_report.xlsx",
  );
}

export async function exportTeamPdf(
  params?: ReportFilters,
) {
  const response = await api.get(
    "/reports/teams/export/pdf",
    {
      params,
      responseType: "blob",
    },
  );

  downloadBlob(
    response.data,
    "team_report.pdf",
  );
}

export async function exportTeamExcel(
  params?: ReportFilters,
) {
  const response = await api.get(
    "/reports/teams/export/excel",
    {
      params,
      responseType: "blob",
    },
  );

  downloadBlob(
    response.data,
    "team_report.xlsx",
  );
}

export async function exportAuditPdf(
  params?: ReportFilters,
) {
  const response = await api.get(
    "/reports/audit/export/pdf",
    {
      params,
      responseType: "blob",
    },
  );

  downloadBlob(
    response.data,
    "audit_report.pdf",
  );
}

export async function exportAuditExcel(
  params?: ReportFilters,
) {
  const response = await api.get(
    "/reports/audit/export/excel",
    {
      params,
      responseType: "blob",
    },
  );

  downloadBlob(
    response.data,
    "audit_report.xlsx",
  );
}